chain/index/events.go

```diff
@@ -0,0 +1,584 @@
+package index
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/ipfs/go-cid"
+	ipld "github.com/ipfs/go-ipld-format"
+	cbg "github.com/whyrusleeping/cbor-gen"
+	"golang.org/x/xerrors"
+
+	"github.com/filecoin-project/go-address"
+	amt4 "github.com/filecoin-project/go-amt-ipld/v4"
+	"github.com/filecoin-project/go-state-types/abi"
+	blockadt "github.com/filecoin-project/specs-actors/actors/util/adt"
+
+	"github.com/filecoin-project/lotus/chain/types"
+)
+
+const maxLookBackForWait = 120 // one hour of tipsets
+
+type executedMessage struct {
+	msg types.ChainMsg
+	rct types.MessageReceipt
+	// events extracted from receipt
+	evs []types.Event
+}
+
+// events are indexed against their inclusion/message tipset when we get the corresponding execution tipset
+func (si *SqliteIndexer) indexEvents(ctx context.Context, tx *sql.Tx, msgTs *types.TipSet, executionTs *types.TipSet) error {
+	if si.actorToDelegatedAddresFunc == nil {
+		return xerrors.Errorf("indexer can not index events without an address resolver")
+	}
+	if si.executedMessagesLoaderFunc == nil {
+		return xerrors.Errorf("indexer can not index events without an event loader")
+	}
+
+	// check if we have an event indexed for any message in the `msgTs` tipset -> if so, there's nothig to do here
+	// this makes event inserts idempotent
+	msgTsKeyCidBytes, err := toTipsetKeyCidBytes(msgTs)
+	if err != nil {
+		return xerrors.Errorf("failed to get tipset key cid: %w", err)
+	}
+
+	// if we've already indexed events for this tipset, mark them as unreverted and return
+	res, err := tx.Stmt(si.stmts.updateEventsToNonRevertedStmt).ExecContext(ctx, msgTsKeyCidBytes)
+	if err != nil {
+		return xerrors.Errorf("failed to unrevert events for tipset: %w", err)
+	}
+	rows, err := res.RowsAffected()
+	if err != nil {
+		return xerrors.Errorf("failed to get rows affected by unreverting events for tipset: %w", err)
+	}
+	if rows > 0 {
+		log.Infof("unreverted %d events for tipset: %s", rows, msgTs.Key())
```

Comment:
```suggestion
		log.Debugf("unreverted %d events for tipset: %s", rows, msgTs.Key())
```

this is the noisiest thing in the logs from running this, it's currently making up 2.9% of my log file, so it happens quite frequently.

Reply:
This is done.